sklearn/preprocessing/discretization.py

```diff
@@ -0,0 +1,261 @@
+# Author: Henry Lin <hlin117@gmail.com>
+
+from __future__ import division, absolute_import
+
+import numbers
+import numpy as np
+import warnings
+
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.preprocessing.data import _transform_selected
+from sklearn.utils.validation import (
+    check_array,
+    check_is_fitted,
+    column_or_1d,
+    FLOAT_DTYPES
+)
+
+
+class KBinsDiscretizer(BaseEstimator, TransformerMixin):
+    """Bins continuous data into k equal width intervals.
+
+    Parameters
+    ----------
+    n_bins : int or array-like, shape (n_features,) (default=2)
+        The number of bins to produce. The intervals for the bins are
+        determined by the minimum and maximum of the input data.
+        Raises ValueError if ``n_bins < 2``.
+
+        If ``n_bins`` is an array, and there is an ignored feature at
+        index ``i``, ``n_bins[i]`` will be ignored.
+
+    ignored_features : int array-like (default=None)
+        Column indices of ignored features. (Example: Categorical features.)
+        If ``None``, all features will be discretized.
+
+    Attributes
+    ----------
+    offset_ : float array, shape (n_features,)
+        Minimum value per feature in ``X``. An ignored feature at index
+        ``i`` will have ``offset_[i] == 0``.
+
+    n_bins_ : int array, shape (n_features,)
+        Number of bins per feature. An ignored feature at index ``i``
+        will have ``n_bins_[i] == 0``.
+
+    bin_width_ : float array, shape (n_features,)
+        The width of each bin. Ignored features will have width ``0``.
+
+    transformed_features_ : int array, shape (n_features,)
+        Features which are transformed.
+
+    Example
+    -------
+    >>> X = [[-2, 1, -4,   -1],
+    ...      [-1, 2, -3, -0.5],
+    ...      [ 0, 3, -2,  0.5],
+    ...      [ 1, 4, -1,    2]]
+    >>> est = KBinsDiscretizer(n_bins=3)
+    >>> est.fit(X)  # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
+    KBinsDiscretizer(...)
+    >>> Xt = est.transform(X)
+    >>> Xt  # doctest: +SKIP
+    array([[ 0., 0., 0., 0.],
+           [ 1., 1., 1., 0.],
+           [ 2., 2., 2., 1.],
+           [ 2., 2., 2., 2.]])
+
+    Sometimes it may be useful to convert the data back into the original
+    feature space. The ``inverse_transform`` function converts the binned
+    data into the original feature space. Each value will be a distance
+    of ``bin_width_ / 2`` from the nearest bin edge.
+
```

Comment:
It might be interesting to mention that the `inverse_transform` does not lead to the original data due to discretization rounding. If not in the docstring, I would add it in the User Guide.